app/server/lib/OIDCConfig.ts

```diff
@@ -167,33 +306,58 @@ export class OIDCConfig {
         profile,
       }));
 
-      delete mreq.session.oidc;
+      mreq.session.oidc = {
+        idToken: tokenSet.id_token, // keep idToken for logout
+      };
       res.redirect(targetUrl ?? '/');
     } catch (err) {
       log.error(`OIDC callback failed: ${err.stack}`);
-      // Delete the session data even if the login failed.
+      if (Object.prototype.hasOwnProperty.call(err, 'response')) {
+        log.error(`Response received: ${err.response?.body ?? err.response}`);
+      }
+
+      // Delete entirely the session data when the login failed.
       // This way, we prevent several login attempts.
       //
       // Also session deletion must be done before sending the response.
-      delete mreq.session.oidc;
-      res.status(500).send(`OIDC callback failed.`);
+      try {
+        const mreq = this._getRequestWithSession(req);
+        delete mreq.session.oidc;
+      } catch (e) {
+        // _getRequestWithSession may fail because there is no session,
+        // in such a case, ignore the exception, just log some error.
+        log.warn('Could not retrieve session to delete the oidc protections');
```

Comment:
If `mreq.session` is missing, the error reporting would be confusing: we'd have a misleading message "OIDC callback failed" (in like 314), and then this second message about retrieving session. Maybe better to check for session once at the start, in a separate check? If it's absent, we don't need to worry about cleaning it either.